### PR TITLE
ceph-dencoder: OSD - Add missing types

### PR DESCRIPTION
**Pull request by NitzanMordhai**
Currently, ceph-dencoder lacks certain osd type, preventing us from accurately checking the ceph corpus for encode-decode mismatches. This pull request aims to address this issue by adding the missing types to ceph-dencoder.

To successfully incorporate these types into ceph-dencoder, we need to introduce the necessary `dump` and `generate_test_instances` functions that was missing in some types. These functions are essential for proper encode and decode of the added types.

This PR will enhance the functionality of ceph-dencoder by including the missing types, enabling a comprehensive analysis of encode-decode consistency. With the addition of these types, we can ensure the robustness and correctness of the ceph corpus.

This update will significantly contribute to improving the overall reliability and accuracy of ceph-dencoder. It allows for a more comprehensive assessment of the encode-decode behavior, leading to enhanced data integrity and stability within the ceph ecosystem.

Fixes: https://tracker.ceph.com/issues/61788

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/test/encoding/readble.py
Comment (rzarzynski) [inline src/test/encoding/readble.py:1]: This doesn't look like a mds- or osd-targetting change. Maybe we have it (+ directly related parts) in a separated commit?

File: src/test/encoding/readble.py
Comment (rzarzynski) [inline src/test/encoding/readble.py:158]: Who's going  to execute `readable.py`? I can't find a caller in this PR.

File: src/tools/ceph-dencoder/osd_types.h
Comment (rzarzynski) [inline src/tools/ceph-dencoder/osd_types.h:74]: nit: typo.

File: src/tools/ceph-dencoder/osd_types.h
Comment (rzarzynski) [inline src/tools/ceph-dencoder/osd_types.h:3]: How about decoupling OSD from MDS? Now this PR needs reviews from 2 components.

File: src/test/encoding/readble.py
Comment (NitzanMordhai) [reply src/test/encoding/readble.py:1]: thanks, need to delete it from that commit

File: src/tools/ceph-dencoder/osd_types.h
Comment (NitzanMordhai) [reply src/tools/ceph-dencoder/osd_types.h:3]: will do that, i thought MDS changes won't be too much, but now it is reasonable to decoupling them

File: src/tools/ceph-dencoder/osd_types.h
Comment (NitzanMordhai) [reply src/tools/ceph-dencoder/osd_types.h:74]: Done

File: src/osd/osd_types.h
Comment (Matan-B) [inline src/osd/osd_types.h:4865]: nit extra space

File: src/tools/ceph-dencoder/denc_plugin.h
Comment (Matan-B) [inline src/tools/ceph-dencoder/denc_plugin.h:69]: Does this definition have any users?

File: src/mgr/OSDPerfMetricTypes.h
Comment (Matan-B) [inline src/mgr/OSDPerfMetricTypes.h:9]: Why do we need to include Formatter here?

File: src/mon/CreatingPGs.h
Comment (Matan-B) [inline src/mon/CreatingPGs.h:69]: ?

File: src/osd/SnapMapper.h
Comment (Matan-B) [inline src/osd/SnapMapper.h:166]: SnapMapper support was already useful for my local branches :+1:

File: src/osd/osd_types.cc
Comment (Matan-B) [inline src/osd/osd_types.cc:3742]: nit

File: src/osd/osd_types.cc
Comment (Matan-B) [inline src/osd/osd_types.cc:3754]: nit extra space

File: src/osd/osd_types.cc
Comment (Matan-B) [inline src/osd/osd_types.cc:3806]: nit extra new line

File: src/tools/ceph-dencoder/osd_types.h
Comment (Matan-B) [inline src/tools/ceph-dencoder/osd_types.h:30]: Why do we need both the namespace alias and the original namespaces declared?
I noticed that this was the case also with `ceph::os::Transaction`, `MgrMap::StandbyInfo`, etc.

**Review by Matan-B — COMMENTED** [review]
Overall looks good (and useful), left a few comments

Do we have a test that iterates through all the dencoder supported types? Might be useful before merging

**Comment by Matan-B** [comment]
#52623 #52482 #52481 #52210 are all sharing (roughly) the same title.
May I suggest the following title format?
ceph-dencoder: < component > - Add missing types

File: src/osd/osd_types.h
Comment (NitzanMordhai) [reply src/osd/osd_types.h:4865]: Done

File: src/tools/ceph-dencoder/denc_plugin.h
Comment (NitzanMordhai) [reply src/tools/ceph-dencoder/denc_plugin.h:69]: Not yet, but for future use with arguments

File: src/mgr/OSDPerfMetricTypes.h
Comment (NitzanMordhai) [reply src/mgr/OSDPerfMetricTypes.h:9]: moving it to .cc file

File: src/mon/CreatingPGs.h
Comment (NitzanMordhai) [reply src/mon/CreatingPGs.h:69]: yep.. removing

File: src/osd/osd_types.cc
Comment (NitzanMordhai) [reply src/osd/osd_types.cc:3742]: done

File: src/osd/osd_types.cc
Comment (NitzanMordhai) [reply src/osd/osd_types.cc:3754]: done

File: src/osd/osd_types.cc
Comment (NitzanMordhai) [reply src/osd/osd_types.cc:3806]: done

File: src/tools/ceph-dencoder/osd_types.h
Comment (NitzanMordhai) [reply src/tools/ceph-dencoder/osd_types.h:30]: ceph-dencoder doesn't know how to find type when it doesn't have its namespace, and since we dumped the type without namespace we will "lose" checks for older version, but yes, i'll change that to skip those types, but will add dump for the full namespace

**Comment by NitzanMordhai** [comment]
> Overall looks good (and useful), left a few comments
> 
> Do we have a test that iterates through all the dencoder supported types? Might be useful before merging

Running it for each PR change ../src/test/encoding/check-generated.sh

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
check-generated.sh is failing with few `failed to decode`

```
error: void MonCommand::decode(ceph::buffer::list::const_iterator &) decode past end of struct encoding: Malformed input [buffer:3]
        /home/jenkins-build/build/workspace/ceph-pull-requests/ceph-object-corpus/archive/0.80-rc1-35-g4812150/objects/SnapMapper::object_snaps
**** failed to decode /home/jenkins-build/build/workspace/ceph-pull-requests/ceph-object-corpus/archive/0.80-rc1-35-g4812150/objects/MonCommand/6cde3066e02a9ab7fce3758724e719bb ****
```

**Comment by NitzanMordhai** [comment]
jenkins test make check

File: src/mgr/MetricTypes.h
Comment (rzarzynski) [inline src/mgr/MetricTypes.h:36]: C++17's structured bindings could useful to provide names for the `std::pair`'s members (`query` & `report`?).

File: src/cls/fifo/cls_fifo.cc
Comment (rzarzynski) [inline src/cls/fifo/cls_fifo.cc:19]: nit: should be sorted alphabetically (`c` precedes `i`).

File: src/mgr/DaemonHealthMetric.h
Comment (rzarzynski) [inline src/mgr/DaemonHealthMetric.h:9]: nit: should be sorted alphabetically.

File: src/mgr/OSDPerfMetricTypes.cc
Comment (rzarzynski) [inline src/mgr/OSDPerfMetricTypes.cc:5]: Our includes should appear after any standard ones. I know it's non-new but it would be really nice if we can fix that.

File: src/mgr/OSDPerfMetricTypes.h
Comment (rzarzynski) [inline src/mgr/OSDPerfMetricTypes.h:9]: nit: should be sorted alphabetically.

File: src/mon/CreatingPGs.h
Comment (rzarzynski) [inline src/mon/CreatingPGs.h:99]: ACK.

File: src/mon/MgrMap.h
Comment (rzarzynski) [inline src/mon/MgrMap.h:241]: nit: `const auto&`.

File: src/mon/MonCommand.h
Comment (rzarzynski) [inline src/mon/MonCommand.h:71]: https://www.techopedia.com/definition/27997/quux

File: src/mon/MonMap.cc
Comment (rzarzynski) [inline src/mon/MonMap.cc:120]: nit: `const auto&`.

File: src/mon/MonMap.cc
Comment (rzarzynski) [inline src/mon/MonMap.cc:121]: `std::pair` makes easy to hurt readability.

File: src/mon/MonCommand.h
Comment (rzarzynski) [inline src/mon/MonCommand.h:18]: nit: nit: should be sorted alphabetically.

File: src/os/Transaction.h
Comment (rzarzynski) [inline src/os/Transaction.h:1296]: Why this is needed? Just one line below we still have `Transaction::TransactionData`.

File: src/osd/osd_types.cc
Comment (rzarzynski) [inline src/osd/osd_types.cc:3719]: What's the reason behind stripping these test instances?

File: src/osd/osd_types.h
Comment (rzarzynski) [inline src/osd/osd_types.h:4713]: nit: do we need this blank?

File: src/tools/ceph-dencoder/osd_types.h
Comment (rzarzynski) [inline src/tools/ceph-dencoder/osd_types.h:73]: ?

File: src/tools/ceph-dencoder/osd_types.h
Comment (rzarzynski) [inline src/tools/ceph-dencoder/osd_types.h:75]: nit: a doubled newline.

**Review by rzarzynski — COMMENTED** [review]
Overall LGTM! Just nits and few questions.

File: src/osd/osd_types.h
Comment (NitzanMordhai) [reply src/osd/osd_types.h:4713]: Done, please continue review https://github.com/ceph/ceph/pull/52871

File: src/tools/ceph-dencoder/osd_types.h
Comment (NitzanMordhai) [reply src/tools/ceph-dencoder/osd_types.h:73]: i left it there just as a note, we currently can't include those types.

File: src/tools/ceph-dencoder/osd_types.h
Comment (NitzanMordhai) [reply src/tools/ceph-dencoder/osd_types.h:75]: Done, please continue review https://github.com/ceph/ceph/pull/52871

File: src/osd/osd_types.cc
Comment (NitzanMordhai) [reply src/osd/osd_types.cc:3719]: i missed that change, hit some segfault and didn't place it back, added it back

File: src/cls/fifo/cls_fifo.cc
Comment (NitzanMordhai) [reply src/cls/fifo/cls_fifo.cc:19]: Done

**Pull request closed without merging**
